Replace use of old APIs which are deprecated in OpenSSL 3

This patch updates regression unittests.

diff --git a/regress/unittests/sshbuf/test_sshbuf_getput_crypto.c b/regress/unittests/sshbuf/test_sshbuf_getput_crypto.c
index e3620e9..5ac2ea1 100644
--- a/regress/unittests/sshbuf/test_sshbuf_getput_crypto.c
+++ b/regress/unittests/sshbuf/test_sshbuf_getput_crypto.c
@@ -66,7 +66,7 @@ sshbuf_getput_crypto_tests(void)
 		0xc8, 0xf9, 0xa3, 0x5e, 0x42, 0xbd, 0xd0, 0x47,
 		0x55, 0x0f, 0x69, 0xd8, 0x0e, 0xc2, 0x3c, 0xd4
 	};
-	EC_KEY *eck;
+	EC_GROUP *ecg;
 	EC_POINT *ecp;
 #endif
 	int r;
@@ -224,51 +224,52 @@ sshbuf_getput_crypto_tests(void)
 
 #if defined(OPENSSL_HAS_ECC) && defined(OPENSSL_HAS_NISTP256)
 	TEST_START("sshbuf_put_ec");
-	eck = EC_KEY_new_by_curve_name(ec256_nid);
-	ASSERT_PTR_NE(eck, NULL);
-	ecp = EC_POINT_new(EC_KEY_get0_group(eck));
+	ecg = EC_GROUP_new_by_curve_name(ec256_nid);
+	ASSERT_PTR_NE(ecg, NULL);
+	ecp = EC_POINT_new(ecg);
 	ASSERT_PTR_NE(ecp, NULL);
 	MKBN(ec256_x, bn_x);
 	MKBN(ec256_y, bn_y);
-	ASSERT_INT_EQ(EC_POINT_set_affine_coordinates_GFp(
-	    EC_KEY_get0_group(eck), ecp, bn_x, bn_y, NULL), 1);
-	ASSERT_INT_EQ(EC_KEY_set_public_key(eck, ecp), 1);
+	ASSERT_INT_EQ(EC_POINT_set_affine_coordinates(
+	    ecg, ecp, bn_x, bn_y, NULL), 1);
 	BN_free(bn_x);
 	BN_free(bn_y);
-	EC_POINT_free(ecp);
 	p1 = sshbuf_new();
 	ASSERT_PTR_NE(p1, NULL);
-	ASSERT_INT_EQ(sshbuf_put_eckey(p1, eck), 0);
+	ASSERT_INT_EQ(sshbuf_put_ec(p1, ecp, ecg), 0);
 	ASSERT_INT_EQ(sshbuf_get_string_direct(p1, &d, &s), 0);
 	ASSERT_SIZE_T_EQ(s, sizeof(expec256));
 	ASSERT_MEM_EQ(d, expec256, sizeof(expec256));
 	sshbuf_free(p1);
-	EC_KEY_free(eck);
+	EC_POINT_free(ecp);
+	EC_GROUP_free(ecg);
 	TEST_DONE();
 
 	TEST_START("sshbuf_get_ec");
-	eck = EC_KEY_new_by_curve_name(ec256_nid);
-	ASSERT_PTR_NE(eck, NULL);
+	ecg = EC_GROUP_new_by_curve_name(ec256_nid);
+	ASSERT_PTR_NE(ecg, NULL);
+	ecp = EC_POINT_new(ecg);
+	ASSERT_PTR_NE(ecp, NULL);
 	p1 = sshbuf_new();
 	ASSERT_PTR_NE(p1, NULL);
 	ASSERT_INT_EQ(sshbuf_put_string(p1, expec256, sizeof(expec256)), 0);
 	ASSERT_SIZE_T_EQ(sshbuf_len(p1), sizeof(expec256) + 4);
 	ASSERT_INT_EQ(sshbuf_put_u8(p1, 0x00), 0);
-	ASSERT_INT_EQ(sshbuf_get_eckey(p1, eck), 0);
+	ASSERT_INT_EQ(sshbuf_get_ec(p1, ecp, ecg), 0);
 	bn_x = BN_new();
 	bn_y = BN_new();
 	ASSERT_PTR_NE(bn_x, NULL);
 	ASSERT_PTR_NE(bn_y, NULL);
-	ASSERT_INT_EQ(EC_POINT_get_affine_coordinates_GFp(
-	    EC_KEY_get0_group(eck), EC_KEY_get0_public_key(eck),
-	    bn_x, bn_y, NULL), 1);
+	ASSERT_INT_EQ(EC_POINT_get_affine_coordinates(
+	    ecg, ecp, bn_x, bn_y, NULL), 1);
 	MKBN(ec256_x, bn);
 	MKBN(ec256_y, bn2);
 	ASSERT_INT_EQ(BN_cmp(bn_x, bn), 0);
 	ASSERT_INT_EQ(BN_cmp(bn_y, bn2), 0);
 	ASSERT_SIZE_T_EQ(sshbuf_len(p1), 1);
 	sshbuf_free(p1);
-	EC_KEY_free(eck);
+	EC_POINT_free(ecp);
+	EC_GROUP_free(ecg);
 	BN_free(bn_x);
 	BN_free(bn_y);
 	BN_free(bn);
diff --git a/regress/unittests/sshbuf/test_sshbuf_getput_fuzz.c b/regress/unittests/sshbuf/test_sshbuf_getput_fuzz.c
index 3b48958..ff1c753 100644
--- a/regress/unittests/sshbuf/test_sshbuf_getput_fuzz.c
+++ b/regress/unittests/sshbuf/test_sshbuf_getput_fuzz.c
@@ -36,7 +36,8 @@ attempt_parse_blob(u_char *blob, size_t len)
 #ifdef WITH_OPENSSL
 	BIGNUM *bn;
 #if defined(OPENSSL_HAS_ECC) && defined(OPENSSL_HAS_NISTP256)
-	EC_KEY *eck;
+	EC_GROUP *ecg;
+	EC_POINT *ecp;
 #endif /* defined(OPENSSL_HAS_ECC) && defined(OPENSSL_HAS_NISTP256) */
 #endif /* WITH_OPENSSL */
 	u_char *s;
@@ -62,10 +63,13 @@ attempt_parse_blob(u_char *blob, size_t len)
 	sshbuf_get_bignum2(p1, &bn);
 	BN_clear_free(bn);
 #if defined(OPENSSL_HAS_ECC) && defined(OPENSSL_HAS_NISTP256)
-	eck = EC_KEY_new_by_curve_name(NID_X9_62_prime256v1);
-	ASSERT_PTR_NE(eck, NULL);
-	sshbuf_get_eckey(p1, eck);
-	EC_KEY_free(eck);
+	ecg = EC_GROUP_new_by_curve_name(NID_X9_62_prime256v1);
+	ASSERT_PTR_NE(ecg, NULL);
+	ecp = EC_POINT_new(ecg);
+	ASSERT_PTR_NE(ecp, NULL);
+	sshbuf_get_ec(p1, ecp, ecg);
+	EC_POINT_free(ecp);
+	EC_GROUP_free(ecg);
 #endif /* defined(OPENSSL_HAS_ECC) && defined(OPENSSL_HAS_NISTP256) */
 #endif /* WITH_OPENSSL */
 	sshbuf_free(p1);
diff --git a/sshbuf-getput-crypto.c b/sshbuf-getput-crypto.c
index e7bffe2..92291c4 100644
--- a/sshbuf-getput-crypto.c
+++ b/sshbuf-getput-crypto.c
@@ -90,41 +90,6 @@ sshbuf_get_ec(struct sshbuf *buf, EC_POINT *v, const EC_GROUP *g)
 	}
 	return 0;
 }
-
-int
-sshbuf_get_eckey(struct sshbuf *buf, EC_KEY *v)
-{
-	EC_POINT *pt = EC_POINT_new(EC_KEY_get0_group(v));
-	int r;
-	const u_char *d;
-	size_t len;
-
-	if (pt == NULL) {
-		SSHBUF_DBG(("SSH_ERR_ALLOC_FAIL"));
-		return SSH_ERR_ALLOC_FAIL;
-	}
-	if ((r = sshbuf_peek_string_direct(buf, &d, &len)) < 0) {
-		EC_POINT_free(pt);
-		return r;
-	}
-	if ((r = get_ec(d, len, pt, EC_KEY_get0_group(v))) != 0) {
-		EC_POINT_free(pt);
-		return r;
-	}
-	if (EC_KEY_set_public_key(v, pt) != 1) {
-		EC_POINT_free(pt);
-		return SSH_ERR_ALLOC_FAIL; /* XXX assumption */
-	}
-	EC_POINT_free(pt);
-	/* Skip string */
-	if (sshbuf_get_string_direct(buf, NULL, NULL) != 0) {
-		/* Shouldn't happen */
-		SSHBUF_DBG(("SSH_ERR_INTERNAL_ERROR"));
-		SSHBUF_ABORT();
-		return SSH_ERR_INTERNAL_ERROR;
-	}
-	return 0;
-}
 #endif /* OPENSSL_HAS_ECC */
 
 int
@@ -170,21 +135,33 @@ sshbuf_put_ec(struct sshbuf *buf, const EC_POINT *v, const EC_GROUP *g)
 	return ret;
 }
 
-int
-sshbuf_put_eckey(struct sshbuf *buf, const EC_KEY *v)
-{
-	return sshbuf_put_ec(buf, EC_KEY_get0_public_key(v),
-	    EC_KEY_get0_group(v));
-}
-
 int
 sshbuf_put_ec_pkey(struct sshbuf *buf, EVP_PKEY *pkey)
 {
-	const EC_KEY *ec;
+	EC_POINT *p = NULL;
+	EC_GROUP *g = NULL;
+	u_char *dg = NULL;
+	const u_char *pdg = NULL;
+	u_char *dp = NULL;
+	int glen, plen;
+	int ret;
 
-	if ((ec = EVP_PKEY_get0_EC_KEY(pkey)) == NULL)
-		return SSH_ERR_LIBCRYPTO_ERROR;
-	return sshbuf_put_eckey(buf, ec);
+	if ((glen = i2d_KeyParams(pkey, &dg)) < 0 ||
+	    (pdg = dg) == NULL ||
+	    (g = d2i_ECPKParameters(NULL, &pdg, glen)) == NULL ||
+	    (plen = EVP_PKEY_get1_encoded_public_key(pkey, &dp)) == 0 ||
+	    (p = EC_POINT_new(g)) == NULL ||
+	    EC_POINT_oct2point(g, p, dp, plen, NULL) == 0) {
+		ret = SSH_ERR_LIBCRYPTO_ERROR;
+		goto out;
+	}
+	ret = sshbuf_put_ec(buf, p, g);
+out:
+	EC_POINT_free(p);
+	OPENSSL_free(dp);
+	EC_GROUP_free(g);
+	OPENSSL_free(dg);
+	return ret;
 }
 #endif /* OPENSSL_HAS_ECC */
 #endif /* WITH_OPENSSL */
diff --git a/sshbuf.h b/sshbuf.h
index 49c32af..67fab2e 100644
--- a/sshbuf.h
+++ b/sshbuf.h
@@ -221,9 +221,7 @@ int	sshbuf_get_bignum2(struct sshbuf *buf, BIGNUM **valp);
 int	sshbuf_put_bignum2(struct sshbuf *buf, const BIGNUM *v);
 # ifdef OPENSSL_HAS_ECC
 int	sshbuf_get_ec(struct sshbuf *buf, EC_POINT *v, const EC_GROUP *g);
-int	sshbuf_get_eckey(struct sshbuf *buf, EC_KEY *v);
 int	sshbuf_put_ec(struct sshbuf *buf, const EC_POINT *v, const EC_GROUP *g);
-int	sshbuf_put_eckey(struct sshbuf *buf, const EC_KEY *v);
 int	sshbuf_put_ec_pkey(struct sshbuf *buf, EVP_PKEY *pkey);
 # endif /* OPENSSL_HAS_ECC */
 #endif /* WITH_OPENSSL */
diff --git a/ssh-ecdsa.c b/ssh-ecdsa.c
index 695ed45..c9c629f 100644
--- a/ssh-ecdsa.c
+++ b/ssh-ecdsa.c
@@ -234,6 +234,7 @@ ssh_ecdsa_deserialize_public(const char *ktype, struct sshbuf *b,
 	int r;
 	char *curve = NULL;
 	EVP_PKEY *pkey = NULL;
+	EC_POINT *pt = NULL;
 	EC_KEY *ec = NULL;
 
 	if ((key->ecdsa_nid = sshkey_ecdsa_nid_from_name(ktype)) == -1)
@@ -244,17 +245,21 @@ ssh_ecdsa_deserialize_public(const char *ktype, struct sshbuf *b,
 		r = SSH_ERR_EC_CURVE_MISMATCH;
 		goto out;
 	}
-	if ((ec = EC_KEY_new_by_curve_name(key->ecdsa_nid)) == NULL) {
+	if ((ec = EC_KEY_new_by_curve_name(key->ecdsa_nid)) == NULL ||
+	    (pt = EC_POINT_new(EC_KEY_get0_group(ec))) == NULL) {
 		r = SSH_ERR_LIBCRYPTO_ERROR;
 		goto out;
 	}
-	if ((r = sshbuf_get_eckey(b, ec)) != 0)
+	if ((r = sshbuf_get_ec(b, pt, EC_KEY_get0_group(ec))) != 0)
 		goto out;
-	if (sshkey_ec_validate_public(EC_KEY_get0_group(ec),
-	    EC_KEY_get0_public_key(ec)) != 0) {
+	if (sshkey_ec_validate_public(EC_KEY_get0_group(ec), pt) != 0) {
 		r = SSH_ERR_KEY_INVALID_EC_VALUE;
 		goto out;
 	}
+	if (EC_KEY_set_public_key(ec, pt) != 1) {
+		r = SSH_ERR_ALLOC_FAIL;
+		goto out;
+	}
 	if ((pkey = EVP_PKEY_new()) == NULL) {
 		r = SSH_ERR_ALLOC_FAIL;
 		goto out;
@@ -274,6 +279,7 @@ ssh_ecdsa_deserialize_public(const char *ktype, struct sshbuf *b,
 	    EC_KEY_get0_public_key(EVP_PKEY_get0_EC_KEY(key->pkey)));
 #endif
  out:
+	EC_POINT_free(pt);
 	EC_KEY_free(ec);
 	EVP_PKEY_free(pkey);
 	free(curve);
